### ISSUE TITLE
Call Graph corrupt and/or incomplete

### ISSUE THREAD

**Issue by DUChost** [doxygen/doxygen#9096]
The call graph is incomplete or/and wrong at least at the marked positions.

I have played around with the different parameters of DoxyGen and also rearranged the C file, unfortunately all without success.

The whole project is in the ZIP file. In the subdirectory "Files" is also the used configuration for DoxyGen.

Best regards UweC

[DOXY_BUG_REPORT.zip](https://github.com/doxygen/doxygen/files/8004782/DOXY_BUG_REPORT.zip)
![call_graph_bug](https://user-images.githubusercontent.com/55077974/152568990-8405f672-2081-4c01-8867-b54a7ee251b4.png)
![ApplicationC](https://user-images.githubusercontent.com/55077974/152569609-a8bb2e98-ef7b-4195-85b9-31743ec2b65e.png)

**Comment by albert-github** [comment]
It looks like the problem is caused by the variable ` usSysTick`
In SRC/Application.h it is defined as a local variable:
```
  UInt16_t usSysTick;
```

and in SRC/Systick_WDT.c as:
```
static volatile UInt16_t usSysTick = 0;
```

In principle these 2 values should not influence each other in the documentation (like they also don't in the compiler).

As a workaround you could redefine local variable in SRC/Application.h to e.g.  `usSysTickLow` or use the setting `EXTRACT_STATIC=YES`

**Comment by DUChost** [comment]
Dear Albert,

yes, using EXTRACT_STATIC works in my case: Thanks!
Do you have any other suggestions for my files and by the way, how did you solve that problem?

--> Did you recognize that "Process_RealTime" is not called by "main" but via callback from "SysTick_WDT"?!

Again, Thanks a lot!

**Comment by albert-github** [comment]
> how did you solve that problem?

Trial and error, i.e. reducing the number of files and after that reducing the remaining files, a bit of luck and some educational guessing / experience.

Note: that when using `EXTRACT_STATIC=YES` and using `SOURCE_BROWSER=YES` in the function `Process` the local variable `usSysTick` is "linked" to the `static volatile` variable! So better to use the`usSysTickLow` workaround.

> --> Did you recognize that "Process_RealTime" is not called by "main" but via callback from "SysTick_WDT"?!

The problem here is that in `Appication.c` in the function `main` we have:
```
SysTick_Init( Process_RealTime );
```
i.e. for doxygn the function `Process_RealTime` is used here, despite the fact that the actual calling takes place in `Systick_WDT.c` where the function is "stored in a variable` and later this `variable` is called. Doxygen is not good in recognizing this kind of usage (actually doesn't recognize the construct). So `CALL_GRAPH` is a bit misleading.

**Comment by DUChost** [comment]
With EXTRACT_STATIC=NO and renamed the static variable in SysTick_WDT to usSysTick_WDT the problem changed but in fact is still present. But with usSysTick_WDT and EXTRACT_STATIC=YES it seems to work correctly. 

Does that mean in general that any name ambguity for some static variable in a project will lead to my initial problem again?

**Comment by albert-github** [comment]
> With EXTRACT_STATIC=NO and renamed the static variable in SysTick_WDT to usSysTick_WDT the problem changed but in fact is still present. But with usSysTick_WDT and EXTRACT_STATIC=YES it seems to work correctly.

Something I overlooked,  (and didn't expect), but indeed there is a difference in between `Process_HighTickRate` and `Process_LowTickRate` depending on the setting of `EXTRACT_STATIC`.

> Does that mean in general that any name ambiguity for some static variable in a project will lead to my initial problem again?

I think / fear that this is indeed the case.

**Comment by DUChost** [comment]
Thanks a lot for your help again!

**Comment by albert-github** [comment]
Please don't close this issue as it incorporates still useful information regarding a still existing bug.
We found a workaround for your case, but the actual problem is still present.

**Comment by septatrix** [comment]
Is there any method to get more verbose output about why this might be the case? Cause we have the same issue for years now

**Comment by albert-github** [comment]
There is,to the best of my knowledge, no real debug information for finding this problem.

**Comment by septatrix** [comment]
Would another code sample help? I also tried running with clang to rule out any parsing errors but clang does not generate any call graphs at all for me (not sure if that is supposed to be implemented yet)

**Comment by albert-github** [comment]
I assume that you mean with "running with clang" you mean that you ran doxygen with `CLANG_ASSISTED_PARSING=YES`, this should give call graphs as well.
- Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.
- Please also specify the **full** doxygen version used (`doxygen -v`).

**Comment by septatrix** [comment]
> I assume that you mean with "running with clang" you mean that you ran doxygen with `CLANG_ASSISTED_PARSING=YES`, this should give call graphs as well.

This was apparently a problem with the doxygen version from my distributions repo and did not happen with the self-compiled version.
 
> * Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

Sure, took me a while to get rid of all the unnecessary (and kinda private) stuff but I now got a [VERY minimal example](https://github.com/doxygen/doxygen/files/8892437/example.tar.gz).
The call graph of `os_exec` is incomplete. Removing the line with the enum fixes it for some reason.

> * Please also specify the **full** doxygen version used (`doxygen -v`).

Tested with doxygen 1.9.1 from fedora 36 and 1.9.4 (with 5198966c8d5fec89116d025c74934ac03ea511fa cherry-picked) from source.

**Comment by albert-github** [comment]
@septatrix 
In your MWE I'm missing the doxygen settings file (Doxyfile).

**Comment by septatrix** [comment]
I am fairly certain it is included...note that the archive does not have a top-level directory:

```
tar tvf example.tar.gz
-rw-r--r-- septatrix/septatrix 116205 2022-06-13 17:21 Doxyfile
drwxr-xr-x septatrix/septatrix      0 2022-06-13 17:16 MWE/
-rw-r--r-- septatrix/septatrix   1015 2022-06-13 17:18 MWE/os_scheduler.h
-rw-r--r-- septatrix/septatrix    577 2022-06-13 17:18 MWE/os_scheduler.c
```

**Comment by albert-github** [comment]
Sorry I overlooked the file a the top level.

**Comment by albert-github** [comment]
Regarding the problem as shown by @septatrix, I've just pushed a proposed patch, pull request #9396

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.5.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
